internal/character/march7th/talent.go

```diff
@@ -0,0 +1,123 @@
+package march7th
+
+import (
+	"github.com/simimpact/srsim/pkg/engine/event"
+	"github.com/simimpact/srsim/pkg/engine/info"
+	"github.com/simimpact/srsim/pkg/engine/modifier"
+	"github.com/simimpact/srsim/pkg/key"
+	"github.com/simimpact/srsim/pkg/model"
+)
+
+const (
+	Talent             = "march7th-talent"
+	TalentCount        = "march7th-talentcount"
+	MarchCounterMark   = "march7th-counter-mark"
+	MarchCounterAttack = "march7th-counter"
+)
+
+type talentState struct {
+	maxCounters float64
+}
+
+func init() {
+	modifier.Register(MarchCounterMark, modifier.Config{
+		Listeners: modifier.Listeners{
+			OnAfterAttack: talentCounterAttack,
+		},
+	})
+
+	modifier.Register(TalentCount, modifier.Config{
+		Stacking: modifier.Replace,
+		Count:    2,
+		MaxCount: 2,
+	})
+
+	// The actual talent modifier
+	modifier.Register(Talent, modifier.Config{
+		Listeners: modifier.Listeners{
+			OnPhase2: func(mod *modifier.Instance) {
+				// Perhaps keep the max count allowed stored in a state of the Talent modifier, to avoid having to grab with CharacterInfo
+				maxCounters := mod.State().(talentState).maxCounters
+				mod.Engine().AddModifier(mod.Owner(), info.Modifier{
+					Name:     TalentCount,
+					Source:   mod.Source(),
+					Count:    maxCounters,
+					MaxCount: maxCounters,
+				})
+			},
+		},
+	})
+}
+
+func (c *char) initTalent() {
+	talentCount := 2.0
+	if c.info.Eidolon >= 4 {
+		talentCount += 1
+	}
+	c.engine.AddModifier(c.id, info.Modifier{
+		Name:   Talent,
+		Source: c.id,
+		State: talentState{
+			maxCounters: talentCount,
+		},
+	})
+	c.engine.AddModifier(c.id, info.Modifier{
+		Name:     TalentCount,
+		Source:   c.id,
+		Count:    talentCount,
+		MaxCount: talentCount,
+	})
+}
+
+func (c *char) applyCounterMark(e event.AttackStart) {
+	hasShieldedTarget := false
+	for _, ally := range c.engine.Characters() {
+		hasShieldedTarget = hasShieldedTarget || c.engine.IsShielded(ally)
+	}
+
+	haveCounters := c.engine.HasModifier(c.id, TalentCount)
```

Comment:
Instead of calling `HasModifier` use `ModifierStackCount` and check if count > 0. There are potential edge cases where the stack count will be 0 but modifier will still be active